c_emulator/riscv_callbacks.cpp

```diff
@@ -1,139 +1,99 @@
+#include "riscv_callbacks_if.h"
 #include "riscv_callbacks.h"
-#include "riscv_config.h"
-#include "riscv_platform_impl.h"
-#include "riscv_sail.h"
-#include <stdlib.h>
-#include <vector>
-#include <inttypes.h>
+#include <set>
 
-void print_lbits_hex(lbits val, int length = 0)
+std::set<callbacks_if *> callbacks;
+
+void register_callback_if(callbacks_if *cbi)
 {
-  if (length == 0) {
-    length = val.len;
-  }
-  std::vector<uint8_t> data(length);
-  mpz_export(data.data(), nullptr, -1, 1, 0, 0, *val.bits);
-  for (int i = length - 1; i >= 0; --i) {
-    fprintf(trace_log, "%02" PRIX8, data[i]);
-  }
-  fprintf(trace_log, "\n");
+  callbacks.insert(cbi);
+}
+
+void remove_callback_if(callbacks_if *cbi)
+{
+  callbacks.erase(cbi);
```

Comment:
```suggestion
  std::erase(callbacks, cbi);
```
This is C++20, otherwise
```suggestion
callbacks.erase(std::remove(callbacks.begin(), callbacks.end(), cbi), vec.end());
```